### ISSUE TITLE
The results of pipeline parallelism cannot be reproduced when using different numbers of partitions

### ISSUE THREAD

**Issue by liulixinkerry** [pytorch/pytorch#55036]
## 🐛 Bug

_The code of bug report is based on [the tutorial on pipeline parallelism](https://pytorch.org/tutorials/intermediate/pipeline_tutorial.html)._

1. When partitioning a transformer into different numbers of partitions, outputs are different.
2. When using different `chunks` of pipe, outputs are different.

## To Reproduce

Used code: 

1. change the `targets.cuda(1)` in [L353](https://github.com/pytorch/tutorials/blob/master/intermediate_source/pipeline_tutorial.py#L353) and [L385](https://github.com/pytorch/tutorials/blob/master/intermediate_source/pipeline_tutorial.py#L385) to `targets.cuda(num_gpus - 1)` 
1. setup the random seed in [L48](https://github.com/pytorch/tutorials/blob/master/intermediate_source/pipeline_tutorial.py#L48)
```import numpy as np
import random
import os

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
os.environ["PYTHONHASHSEED"] = str(seed)
```

1. use different [`num_gpus`](https://github.com/pytorch/tutorials/blob/master/intermediate_source/pipeline_tutorial.py#L277) (such as `1`, `2`, `4`) and output logs are different. However, if we fix the `num_gpus` and run the code twice, the results are reproducible.
2. use different [`chunks`](https://github.com/pytorch/tutorials/blob/master/intermediate_source/pipeline_tutorial.py#L300) and output logs are also different.

```
import sys
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
import tempfile
from torch.nn import TransformerEncoder, TransformerEncoderLayer

import numpy as np
import random
import os

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
os.environ["PYTHONHASHSEED"] = str(seed)

if sys.platform == "win32":
    print("Windows platform is not supported for pipeline parallelism")
    sys.exit(0)
if torch.cuda.device_count() < 2:
    print("Need at least two GPU devices for this tutorial")
    sys.exit(0)


class Encoder(nn.Module):
    def __init__(self, ntoken, ninp, dropout=0.5):
        super(Encoder, self).__init__()
        self.src_mask = None
        self.pos_encoder = PositionalEncoding(ninp, dropout)
        self.encoder = nn.Embedding(ntoken, ninp)
        self.ninp = ninp
        self.init_weights()

    def init_weights(self):
        initrange = 0.1
        self.encoder.weight.data.uniform_(-initrange, initrange)

    def _generate_square_subsequent_mask(self, sz):
        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
        mask = (
            mask.float()
            .masked_fill(mask == 0, float("-inf"))
            .masked_fill(mask == 1, float(0.0))
        )
        return mask

    def forward(self, src):
        if self.src_mask is None or self.src_mask.size(0) != src.size(0):
            device = src.device
            mask = self._generate_square_subsequent_mask(src.size(0)).to(device)
            self.src_mask = mask

        src = self.encoder(src) * math.sqrt(self.ninp)
        return self.pos_encoder(src)


class Decoder(nn.Module):
    def __init__(self, ntoken, ninp):
        super(Decoder, self).__init__()
        self.decoder = nn.Linear(ninp, ntoken)
        self.init_weights()

    def init_weights(self):
        initrange = 0.1
        self.decoder.bias.data.zero_()
        self.decoder.weight.data.uniform_(-initrange, initrange)

    def forward(self, inp):
        return self.decoder(inp)


class PositionalEncoding(nn.Module):
    def __init__(self, d_model, dropout=0.1, max_len=5000):
        super(PositionalEncoding, self).__init__()
        self.dropout = nn.Dropout(p=dropout)

        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(
            torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model)
        )
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        pe = pe.unsqueeze(0).transpose(0, 1)
        self.register_buffer("pe", pe)

    def forward(self, x):
        x = x + self.pe[: x.size(0), :]
        return self.dropout(x)


import io
import torch
from torchtext.utils import download_from_url, extract_archive
from torchtext.data.utils import get_tokenizer
from torchtext.vocab import build_vocab_from_iterator

url = "https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip"
test_filepath, valid_filepath, train_filepath = extract_archive(download_from_url(url))
tokenizer = get_tokenizer("basic_english")
vocab = build_vocab_from_iterator(
    map(tokenizer, iter(io.open(train_filepath, encoding="utf8")))
)


def data_process(raw_text_iter):
    data = [
        torch.tensor([vocab[token] for token in tokenizer(item)], dtype=torch.long)
        for item in raw_text_iter
    ]
    return torch.cat(tuple(filter(lambda t: t.numel() > 0, data)))


train_data = data_process(iter(io.open(train_filepath, encoding="utf8")))
val_data = data_process(iter(io.open(valid_filepath, encoding="utf8")))
test_data = data_process(iter(io.open(test_filepath, encoding="utf8")))

device = torch.device("cuda")


def batchify(data, bsz):
    # Divide the dataset into bsz parts.
    nbatch = data.size(0) // bsz
    # Trim off any extra elements that wouldn't cleanly fit (remainders).
    data = data.narrow(0, 0, nbatch * bsz)
    # Evenly divide the data across the bsz batches.
    data = data.view(bsz, -1).t().contiguous()
    return data.to(device)


batch_size = 20
eval_batch_size = 10
train_data = batchify(train_data, batch_size)
val_data = batchify(val_data, eval_batch_size)
test_data = batchify(test_data, eval_batch_size)


bptt = 35


def get_batch(source, i):
    seq_len = min(bptt, len(source) - 1 - i)
    data = source[i : i + seq_len]
    target = source[i + 1 : i + 1 + seq_len].view(-1)
    return data, target


ntokens = len(vocab.stoi)  # the size of vocabulary
emsize = 4096  # embedding dimension
nhid = 4096  # the dimension of the feedforward network model in nn.TransformerEncoder
nlayers = 12  # the number of nn.TransformerEncoderLayer in nn.TransformerEncoder
nhead = 16  # the number of heads in the multiheadattention models
dropout = 0.2  # the dropout value

from torch.distributed import rpc

tmpfile = tempfile.NamedTemporaryFile()
rpc.init_rpc(
    name="worker",
    rank=0,
    world_size=1,
    rpc_backend_options=rpc.TensorPipeRpcBackendOptions(
        init_method="file://{}".format(tmpfile.name),
        # Specifying _transports and _channels is a workaround and we no longer
        # will have to specify _transports and _channels for PyTorch
        # versions >= 1.8.1
        _transports=["ibv", "uv"],
        _channels=["cuda_ipc", "cuda_basic"],
    ),
)

num_gpus = 2
partition_len = ((nlayers - 1) // num_gpus) + 1

# Add encoder in the beginning.
tmp_list = [Encoder(ntokens, emsize, dropout).cuda(0)]
module_list = []

# Add all the necessary transformer blocks.
for i in range(nlayers):
    transformer_block = TransformerEncoderLayer(emsize, nhead, nhid, dropout)
    if i != 0 and i % (partition_len) == 0:
        module_list.append(nn.Sequential(*tmp_list))
        tmp_list = []
    device = i // (partition_len)
    tmp_list.append(transformer_block.to(device))

# Add decoder in the end.
tmp_list.append(Decoder(ntokens, emsize).cuda(num_gpus - 1))
module_list.append(nn.Sequential(*tmp_list))

from torch.distributed.pipeline.sync import Pipe

# Build the pipeline.
model = Pipe(torch.nn.Sequential(*module_list), chunks=8)


def get_total_params(module: torch.nn.Module):
    total_params = 0
    for param in module.parameters():
        total_params += param.numel()
    return total_params


print("Total parameters in model: {:,}".format(get_total_params(model)))


criterion = nn.CrossEntropyLoss()
lr = 5.0  # learning rate
optimizer = torch.optim.SGD(model.parameters(), lr=lr)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1.0, gamma=0.95)

import time


def train():
    model.train()  # Turn on the train mode
    total_loss = 0.0
    start_time = time.time()
    ntokens = len(vocab.stoi)

    # Train only for 50 batches to keep script execution time low.
    nbatches = min(50 * bptt, train_data.size(0) - 1)

    for batch, i in enumerate(range(0, nbatches, bptt)):
        data, targets = get_batch(train_data, i)
        optimizer.zero_grad()
        # Since the Pipe is only within a single host and process the ``RRef``
        # returned by forward method is local to this node and can simply
        # retrieved via ``RRef.local_value()``.
        output = model(data).local_value()
        # Need to move targets to the device where the output of the
        # pipeline resides.
        loss = criterion(output.view(-1, ntokens), targets.cuda(num_gpus - 1))
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
        optimizer.step()

        total_loss += loss.item()
        log_interval = 1
        if batch % log_interval == 0 and batch > 0:
            cur_loss = total_loss / log_interval
            elapsed = time.time() - start_time
            print(
                "| epoch {:3d} | {:5d}/{:5d} batches | "
                "lr {:02.2f} | ms/batch {:5.2f} | "
                "loss {:5.2f} | ppl {:8.2f}".format(
                    epoch,
                    batch,
                    nbatches // bptt,
                    scheduler.get_lr()[0],
                    elapsed * 1000 / log_interval,
                    cur_loss,
                    math.exp(cur_loss),
                )
            )
            total_loss = 0
            start_time = time.time()


def evaluate(eval_model, data_source):
    eval_model.eval()  # Turn on the evaluation mode
    total_loss = 0.0
    ntokens = len(vocab.stoi)
    # Evaluate only for 50 batches to keep script execution time low.
    nbatches = min(50 * bptt, data_source.size(0) - 1)
    with torch.no_grad():
        for i in range(0, nbatches, bptt):
            data, targets = get_batch(data_source, i)
            output = eval_model(data).local_value()
            output_flat = output.view(-1, ntokens)
            # Need to move targets to the device where the output of the
            # pipeline resides.
            total_loss += (
                len(data) * criterion(output_flat, targets.cuda(num_gpus - 1)).item()
            )
    return total_loss / (len(data_source) - 1)


best_val_loss = float("inf")
epochs = 3  # The number of epochs
best_model = None

for epoch in range(1, epochs + 1):
    epoch_start_time = time.time()
    train()
    val_loss = evaluate(model, val_data)
    print("-" * 89)
    print(
        "| end of epoch {:3d} | time: {:5.2f}s | valid loss {:5.2f} | "
        "valid ppl {:8.2f}".format(
            epoch, (time.time() - epoch_start_time), val_loss, math.exp(val_loss)
        )
    )
    print("-" * 89)

    if val_loss < best_val_loss:
        best_val_loss = val_loss
        best_model = model

    scheduler.step()


test_loss = evaluate(best_model, test_data)
print("=" * 89)
print(
    "| End of training | test loss {:5.2f} | test ppl {:8.2f}".format(
        test_loss, math.exp(test_loss)
    )
)
print("=" * 89)

```
## Expected behavior

Because the sync pipeline parallelism doesn't affect the parameter update, output logs should be the same no matter we modify the `num_gpus` / `chunks` or not.

## Environment

 - PyTorch Version (e.g., 1.8.1):
 - OS (e.g., Linux): Linux
 - How you installed PyTorch: `pip`
 - Python version: 3.8
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: RTX 3090 

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu

**Comment by pritamdamania87** [comment]
I believe the non-determinism is coming from nn.Dropout. For example when you have num_gpus = 2, on GPU 0 you end up with 7 nn.Dropout layers (1 for encoder and 6 for TransformerEncoderLayer), in case of num_gpus = 4 you have 4 nn.Dropout layers (1 for encoder, 3 for TrnasformerEncoderLayer).

For the first microbatch the results will be same, however for the second microbatch in num_gpus = 2 case we've executed 7 nn.Dropouts before but in the num_gpus = 4 case we have only executed 4 nn.Dropouts before on GPU 0 and as a result, the Encoder output for microbatch 1 would be different. The same logic applies when you change the number of chunks.

To validate this theory, I set dropout = 0 and the results are same if you change number of GPUs. Although, we set all the seeds appropriately the non-determinism comes from the fact that how many times the nn.Dropout layer is called.

@liulixinkerry Although there is some non-determinism here I was wondering if you were noticing that this non-determinism actually results in convergence issues? My initial feeling is that over a large training run this shouldn't matter much.

**Comment by pritamdamania87** [comment]
@msbaines @shruti-bh Wondering if you encountered similar reproducibility issues with nn.Dropout while using pipeline parallelism?

**Comment by msbaines** [comment]
Yes, different partitioning will result in slightly different results due to dropout. However, the results for a given partitioning will be deterministic. We did test correctness by verifying that the results were the same under different partitioning when dropout is disabled.

**Comment by tefantasy** [comment]
@pritamdamania87 
Hello, I am working together with @liulixinkerry. I totally agree with your analysis on using nn.Dropout in pipelining, but we've found another problem (unrelated to nn.Dropout) in [this tutorial code](https://pytorch.org/tutorials/intermediate/pipeline_tutorial.html). 

We observed that even with nn.Dropout disabled, changing number of chunks still lead to non-determinism. In fact, the output tensors of forward pass are different, when computed by pipelines with different number of chunks. Here's another piece of code for reproducing this issue:

```python
# Same definitions for Module classes and constants as in the tutorial code.
...

dropout = 0.0

num_gpus = 2
partition_len = ((nlayers - 1) // num_gpus) + 1



# build layers
encoder = Encoder(ntokens, emsize, dropout)
transformer_layers = [TransformerEncoderLayer(emsize, nhead, nhid, dropout) for _ in range(nlayers)]
decoder = Decoder(ntokens, emsize)

encoder1 = Encoder(ntokens, emsize, dropout)
transformer_layers1 = [TransformerEncoderLayer(emsize, nhead, nhid, dropout) for _ in range(nlayers)]
decoder1 = Decoder(ntokens, emsize)

# use the same set of parameters
for src_net, dest_net in zip([encoder, decoder] + transformer_layers, [encoder1, decoder1] + transformer_layers1):
    params_dict = src_net.state_dict()
    params_dict_copy = {key:params_dict[key].clone() for key in params_dict.keys()}
    dest_net.load_state_dict(params_dict_copy)

# build pipeline
net = nn.Sequential(
    nn.Sequential(*([encoder] + transformer_layers[:partition_len])).cuda(0),
    nn.Sequential(*(transformer_layers[partition_len:] + [decoder])).cuda(num_gpus - 1)
)
# Or without pipeline parallelism:
# net = nn.Sequential(*([encoder] + transformer_layers + [decoder])).cuda(0)
net1 = nn.Sequential(
    nn.Sequential(*([encoder1] + transformer_layers1[:partition_len])).cuda(0),
    nn.Sequential(*(transformer_layers1[partition_len:] + [decoder1])).cuda(num_gpus - 1)
)
net = Pipe(net, chunks=2)
net1 = Pipe(net1, chunks=8)

# test using a random vector
random_input = torch.randint(0, 10000, [bptt, batch_size]).cuda(0)
net_out = net(random_input).local_value()
net1_out = net1(random_input).local_value()

print(torch.max(torch.abs(net_out - net1_out)))
```

The output is:
```
tensor(14.0103, device='cuda:1', grad_fn=<MaxBackward1>)
```
\
One thing that might be the reason of this problem is that the input data used in this tutorial is seq-first (shape: seq x batch). However, the pipeline wrapper treats the input tensor as batch-first when splitting microbatches, and also stacks microbatches at dim=0 before output. Therefore, in this case the sequence dimension is actually split and concatenated, by mistake.

After modifying the `Encoder` and `Decoder` to let them accept and output batch-first tensors (and also the data generator functions), the above code outputs:
```
tensor(5.1498e-05, device='cuda:1', grad_fn=<MaxBackward1>)
```

I was wondering if this output is able to prove that pipeline parallelism is now working correctly? The value of 5.1498e-05 might be related to precision issues (my guess), but I am not sure if this value is small enough to be regarded as precision error, in the context of pytorch?

For the original training script in the tutorial, after applying the above modification the training loss is able to remain consistent (across different settings of `chunks`) for the initial ~20 steps, and then divergence could be observed. But there seems to be no problem with training convergence. 

Thanks!

**Comment by pritamdamania87** [comment]
> I was wondering if this output is able to prove that pipeline parallelism is now working correctly? The value of 5.1498e-05 might be related to precision issues (my guess), but I am not sure if this value is small enough to be regarded as an acceptable precision error, in the context of pytorch?

I do believe that should be within acceptable precision error. Thanks for spotting this issue! I'll update the pipeline tutorial to fix this issue.

**Comment by pritamdamania87** [comment]
@liulixinkerry I've fixed the tutorials to handle batch size correctly. Feel free to re-open this issue if there is still a problem that needs to be resolved.
